framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1109,53 +1188,134 @@ int pcm_stop(struct pcm *pcm)
 		return -EINVAL;
 	}
 #ifdef CONFIG_AUDIO_MULTI_SESSION
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
 #else
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
 #endif
 		return oops(pcm, errno, "cannot stop channel");
 
+	/* Remove any pending messages from the message queue */
+	struct timespec st_time;
+	struct audio_msg_s msg;
+	unsigned int size;
+	int prio;
+	while (1) {
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+		if (size != sizeof(msg)) {
+			break;
+		}
+	}
+
 	pcm->prepared = 0;
 	pcm->running = 0;
-	pcm->bufPtr = 0;
-	pcm->nextSize = 0;
-	pcm->nextOffset = 0;
-	pcm->nextBuf = NULL;
+	pcm->draining = 0;
+	pcm->buf_ptr = 0;
+	pcm->next_size = 0;
+	pcm->next_offset = 0;
+	pcm->next_buf = NULL;
 
 	return 0;
 }
 
+/** Stops a PCM. Any data present in the buffers will be dropped.
+ * @param pcm A PCM handle.
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drop(struct pcm *pcm)
+{
+	return pcm_stop(pcm);
+}
+
+/** Stop a PCM preserving pending frames.
+ * @param pcm A PCM handle.
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drain(struct pcm *pcm)
+{
+	if (pcm == NULL) {
+		return -EINVAL;
+	}
+
+	if (!pcm->running) {
+		return oops(pcm, EINVAL, "PCM is already stopped.");
+	}
+
+	if (pcm->flags & PCM_OUT) {
+		struct audio_msg_s msg;
+		unsigned int size;
+		int prio;
+
+		/* Playback case */
+		/* Wait for all enqueued buffers to get dequeued. */
+		while (pcm->buf_ptr > 0) {
+			/* Wait for deque message from kernel */
+			size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
+			if (size != sizeof(msg)) {
+				/* Interrupted by a signal? What to do? */
+				return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
+			}
+			if (msg.msgId == AUDIO_MSG_DEQUEUE) {
+				pcm->buf_ptr--;
+			}
+		}
+
+		return pcm_stop(pcm);
+	} else {
+		/* Recording case */
+		/* Stop the codec */
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		if (ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0) {
+#else
+		if (ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0) {
+#endif
+			return oops(pcm, errno, "cannot stop channel");
+		}
+
+		pcm->prepared = 0;
+		pcm->running = 0;
+		pcm->buf_ptr = 0;
+		/* We have stopped the codec, but the user can still call read API to
+		   get the data that has already been recorded */
+		pcm->draining = 1;
+
+		return 0;
```

Comment:
As I Know, when user call during the record, it should reveal how many frames skipped.
Could you please check it & modify return type?

Reply:
I made this implementation according to the alsa-project website API reference page.
https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#ga49afc5b8527f30c33fafa476533c9f86

It says below for the pcm_drain() API:
Stop a PCM preserving pending frames. 
For playback wait for all pending frames to be played and then stop the PCM. For capture stop PCM permitting to retrieve residual frames.

It does not say about pending / skipped frames. 
Please let me know your opinion.